src/core/document.js

```diff
@@ -1711,19 +1711,43 @@ class PDFDocument {
       : clearGlobalCaches();
   }
 
-  async #collectFieldObjects(name, fieldRef, promises, annotationGlobals) {
+  async #collectFieldObjects(
+    name,
+    fieldRef,
+    promises,
+    annotationGlobals,
+    visitedRefs
+  ) {
     const { xref } = this;
 
-    if (!(fieldRef instanceof Ref)) {
+    if (!(fieldRef instanceof Ref) || visitedRefs.has(fieldRef)) {
       return;
     }
+    visitedRefs.put(fieldRef);
     const field = await xref.fetchAsync(fieldRef);
     if (!(field instanceof Dict)) {
       return;
     }
     if (field.has("T")) {
       const partName = stringToPDFString(await field.getAsync("T"));
       name = name === "" ? partName : `${name}.${partName}`;
+    } else {
+      let obj = field;
+      while (true) {
+        const objRef = obj.getRaw("Parent");
+        if (objRef instanceof Ref && visitedRefs.has(objRef)) {
+          break;
+        }
+        obj = await this.xref.fetchIfRefAsync(objRef);
```

Comment:
We can slightly reduce the amount of "re-checking" needed when handling References.
```suggestion
        obj = obj.getRaw("Parent");
        if (obj instanceof Ref) {
          if (visitedRefs.has(objRef)) {
            break;
          }
          obj = await xref.fetchAsync(objRef);
        }
```